### PR TITLE
Crescent caster

### PR DESCRIPTION
**Pull request by ThatDamnPineapple**
This PR adds the crescent caster, a new dungeon enemy, as well as making some slight changes to the NPC barrier system, using the last nonzero max barrier for barrier bar drawing instead of current maxbarrier.

Follows this ticket: https://trello.com/c/ClVkK0GV/256-cloudstrike-3

Only thing left to do is an actual bestiary entry, I used a placeholder, but that can be done in its own PR, along with other bestiary entries.

### CODE REVIEW

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+using Terraria.GameContent.Bestiary;
+using Terraria.GameContent.ItemDropRules;
+using static Terraria.ModLoader.ModContent;
+using Terraria.GameContent.Bestiary;
+using Terraria.ModLoader.Utilities;
+
+
+namespace StarlightRiver.Content.NPCs.Dungeon
+{
+    public class CrescentCasterBolt
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:36]: This feels like it should be placed at the bottom of the file as opposed to the top, or a seperate file altogether. In general lets try to keep the top class of a file the namesake of the file.

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        }
+
+        public void Render(Effect effect)
+        {
+            trail?.Render(effect);
+            trail2?.Render(effect);
+        }
+    }
+
+    internal class CrescentCaster : ModNPC,IDrawPrimitive
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:123]: lets make sure we have a space after the comma when we use multi-inheritence

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    {
+        public override string Texture => AssetDirectory.DungeonNPC + Name;
+
+        private const int XFRAMES = 3;
+        private readonly float ACCELERATION = 0.15f;
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:128]: should these be constant? The capitilzation scheme seems to imply it but they lack the const keyword

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+        private readonly float MAXSPEED = 2;
+
+        private readonly int aiCounterReset = 800;
+
+        private int xFrame = 0;
+        private int yFrame = 0;
+
+        private int aiCounter = 0;
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:136]: this should probably be a ref property to an ai field

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+        public override void SetBestiary(BestiaryDatabase database, BestiaryEntry bestiaryEntry)
+        {
+            bestiaryEntry.Info.AddRange(new IBestiaryInfoElement[]
+            {
+                BestiaryDatabaseNPCsPopulator.CommonTags.SpawnConditions.Biomes.TheDungeon,
+                new FlavorTextBestiaryInfoElement("[PH] He be castin")
+            });
+        }
+
+        public override void AI()
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:190]: Really nice encapsulation of behaviors here.

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+            NPC.TargetClosest(true);
+            aiCounter++;
+
+            if (Casting)
+                CastBehavior();
+            else
+                WalkBehavior();
+        }
+
+        public override float SpawnChance(NPCSpawnInfo spawnInfo)
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:201]: Hm, I feel like the methods you call from AI should be defined after AI to keep behavior-related methods grouped together.

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+            {
+                case 0: //first frame of casting
+                    xFrame = 1;
+                    NPC.frameCounter = 0;
+                    yFrame = 0;
+                    break;
+                case 1: //Winding up to cast
+                    NPC.frameCounter++;
+
+                    if (NPC.frameCounter % 4 == 0)
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:309]: single line conditionals should be inlined

File: Content/NPCs/Dungeon/CrescentCaster.cs [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+                    starOpacity += 0.1f;
+            }
+            else if (starOpacity > 0)
+                starOpacity -= 0.1f;
+
+            foreach (NPC supportTarget in supportTargets)
+            {
+                BarrierNPC barrierNPC = supportTarget.GetGlobalNPC<BarrierNPC>();
+
+                barrierNPC.MaxBarrier = 250;
```
Comment (ScalarVector1) [inline Content/NPCs/Dungeon/CrescentCaster.cs:395]: should we be hard setting to 250? what if an enemy already has max barrier, shouldn't we be adding?

**Review by ScalarVector1 — COMMENTED** [review]
Overall seems very good, some minor organizational and formatting things should probably be changed for general cleanliness.

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 8 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 7 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs:151]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 6 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 5 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs:103]: thanks

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 4 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 3 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 2 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

File: Content/NPCs/Dungeon/CrescentCaster.cs (hunk 1 above)
Comment (ThatDamnPineapple) [reply Content/NPCs/Dungeon/CrescentCaster.cs]: fixed

**Pull request merged**